pkg/network/bridge/plugin.go

```diff
@@ -237,107 +236,138 @@ func (nbdp *NetworkBridgeDevicePlugin) attachPods() {
 	if err != nil {
 		return err
 	}
 
-	// create the virtual interface that should be connected to the bridge
-	link := &netlink.Veth{
-		LinkAttrs: netlink.LinkAttrs{
-			Name:        linkName,
-			MasterIndex: bridge.Attrs().Index,
-			MTU:         bridge.Attrs().MTU,
-			NetNsID:     bridge.Attrs().NetNsID},
-		PeerName: nicName}
+	// in case of a failure, try clean up by removing container interface
+	defer func(contNs netns.NsHandle) {
+		if !success {
+			inNetnsDo(contNs, func(netns.NsHandle) error {
+				contIface, err := netlink.LinkByName(contIfaceName)
+				if err == nil {
+					netlink.LinkDel(contIface)
+				}
+				hostIface, err := netlink.LinkByName(hostIfaceName)
+				if err == nil {
+					netlink.LinkDel(hostIface)
+				}
+				return nil
+			})
+			hostIface, err := netlink.LinkByName(hostIfaceName)
+			if err == nil {
+				netlink.LinkDel(hostIface)
+			}
+		}
+	}(contNs)
+
+	// enter container namespace and configure veth pair
+	err = inNetnsDo(contNs, func(hostNs netns.NsHandle) error {
+		// create the virtual interface that should be connected to the bridge
+		contIface := &netlink.Veth{
+			LinkAttrs: netlink.LinkAttrs{
+				Name: contIfaceName,
+				MTU:  bridgeIface.Attrs().MTU,
+			},
+			PeerName: hostIfaceName,
+		}
+		err = netlink.LinkAdd(contIface)
+		if err != nil {
+			return err
+		}
+
+		// set interface up
+		err = netlink.LinkSetUp(contIface)
+		if err != nil {
+			return err
+		}
+
+		// get the host side interface
+		hostIface, err := netlink.LinkByName(hostIfaceName)
+		if err != nil {
+			return err
+		}
+
+		// move host side to host namespace
+		err = netlink.LinkSetNsPid(hostIface, int(hostNs))
+		if err != nil {
+			return err
+		}
 
-	err = netlink.LinkAdd(link)
+		return nil
+	})
 	if err != nil {
 		return err
 	}
 
-	// set interface up
-	err = netlink.LinkSetUp(link)
+	// get the host interface back, now from the new namespace
+	hostIface, err := netlink.LinkByName(hostIfaceName)
 	if err != nil {
-		netlink.LinkDel(link)
 		return err
 	}
 
-	// get the peer interface
-	peer, err := netlink.LinkByName(nicName)
+	// set MTU on the host interface
+	err = netlink.LinkSetMTU(hostIface, bridgeIface.Attrs().MTU)
 	if err != nil {
-		netlink.LinkDel(link)
 		return err
 	}
 
-	// add peer to pod namespace
-	err = netlink.LinkSetNsPid(peer, containerPid)
+	// connect host side to bridge
+	err = netlink.LinkSetMaster(hostIface, bridgeIface.(*netlink.Bridge))
 	if err != nil {
-		netlink.LinkDel(link)
 		return err
 	}
 
-	// store current namespace
-	originalNS, err := netns.Get()
+	// set host interface up
+	err = netlink.LinkSetUp(hostIface)
 	if err != nil {
-		netlink.LinkDel(link)
 		return err
 	}
 
-	// get namespace of the pod
-	ns, err := netns.GetFromPid(containerPid)
+	success = true
+
+	return nil
+}
+
+func inNetnsDo(targetNs netns.NsHandle, callback func(originalNs netns.NsHandle) error) error {
+	// store current namespace
+	originalNs, err := netns.Get()
```

Comment:
maybe you need to create a specific namespace to be used by the DS pod on the host side, and not just use whatever you get from netns.Get()?this is the only difference i see from the other examples.

Reply:
There are two other differences.

One is that in other solutions, one side of veth remains in a namespace bound to a temporary one-shot process (CNI plugin), we use long running daemon. That would explain why we remove it only when the DS goes down.

Another difference might be that in some examples, they use named namespaces (in /var/run/netns/).

Reply:
So `virt-handler` is also long-running and it works in this PR https://github.com/kubevirt/kubevirt/pull/1407 flawless. There should also no connection to a process. 

Very unlikely, but did you check if your Pods are stuck in "terminating"? Maybe the pods do not get cleaned up?

Reply:
@rmohr  It took them quite a long time to die, I will check, it sounds likely.

Reply:
They are quite long in "terminating" but they end up removed, still not cleaned up.

Reply:
@phoracek can you try to create a namespace for the host side as well?

Reply:
I will. I just rewrote it using CNI libraries, still not cleaning up.

Reply:
Finally with CNI libraries used, locking and Go 1.10 it works. Tests are raceful, but when tested manually, it works.